File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cs
using System;
using System.Collections.Generic;

class Solution {
    int length;
    int height;
    int[,] node = new int[100, 100]; //최대 100까지
    public int solution(int[,] maps) {
        length = maps.GetLength(0); //map의 x값
        height = maps.GetLength(1); //map의 y값
        //4방향 설정
        int[,] direction = {{1, 0},{0, -1}, {-1, 0}, {0, 1}};
        //현위치 설정
        int[] current_pos = {0, 0};
        int cnt = 0;
        maps[0, 0] = 0; //현 위치 0으로 초기화
        node[0 ,0] = 1; //node의 현재 값 1
        Queue<int[]> que = new Queue<int[]>(); // queue 생성
        while(true)
        {
            for(int i = 0; i < 4; i++) //4방향대로 순환
            {
                int[] next = new int[2]; //다음 x,y 좌표값
                next[0] = current_pos[0] + direction[i,0];
                next[1] = current_pos[1] + direction[i,1];
                
                if(next[0] >= 0 && //x좌표의 최솟값
                   next[0] < length && //x좌표의 최댓값
                   next[1] >= 0 && //y좌표의 최솟값
                   next[1] < height && //y좌표의 최댓값
                   maps[next[0], next[1]] == 1
                  ) 
                {
                    que.Enqueue(next); //다음 좌표를 que에 넣음
                    maps[next[0], next[1]] = 0;
                    node[next[0], next[1]] = node[current_pos[0], current_pos[1]] + 1; //다음 노드는 현재좌표 + 1씩
                    if(next[0] == length - 1 &&
                       next[1] == height - 1)
                    {
                        return node[next[0], next[1]];
                    }

                }
            }
            if(que.Count != 0)
            {
                current_pos = que.Dequeue();    
            }
            else
            {
               break;
            }
        }
        return -1;
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
import java.io.*;

class Solution {
    public int solution(int[][] maps) 
    {   
        int rows = maps.length; //행의 개수
        int cols = maps[0].length; //열의 개수
        
        int[][] node = new int[rows][cols];
        int[][] direct = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}}; //4방향 설정
        
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[] {0, 0});
        node[0][0] = 1; //초기값
        
        while(!queue.isEmpty())
        {
            int[] cur = queue.poll();
            int cx = cur[0];
            int cy = cur[1];
            
            if(cx == rows - 1 && cy == cols - 1)
            {
                return node[cx][cy];
            }
            
            for(int i = 0; i < 4; i++)
            {
                int nx = cx + direct[i][0];
                int ny = cy + direct[i][1];
                
                if(nx >= 0 && ny >= 0 && nx < rows && ny < cols && node[nx][ny] == 0 && maps[nx][ny] == 1)
                {
                    node[nx][ny] = node[cx][cy] + 1;
                    queue.add(new int[] {nx, ny});
                }
            }
        }
        
        return -1;
    }
}